OsmAnd/src/net/osmand/plus/onlinerouting/OnlineRoutingHelper.java

```diff
@@ -104,16 +105,41 @@ public OnlineRoutingResponse calculateRouteOnline(@Nullable String stringKey, @N
 	public OnlineRoutingResponse calculateRouteOnline(@NonNull OnlineRoutingEngine engine, @NonNull List<LatLon> path, @Nullable Float startBearing,
 													  boolean leftSideNavigation, boolean initialCalculation) throws IOException, JSONException {
 		String url = engine.getFullUrl(path, startBearing);
-		String content = makeRequest(url);
+		String method = engine.getHTTPMethod();
+		String body = engine.getRequestBody(path, startBearing);
```

Comment:
Could you explain your reasoning a bit further? [RFC7231](https://www.rfc-editor.org/rfc/rfc7231#section-4.3) explicitly allows request bodies for all HTTP methods except `TRACE`. The default implementation of `engine.getRequestBody` returns `null` anyways, so if the engine does not override `getHTTPMethod` and `getRequestBody`, those two will return `GET` and `null` by default. I think we want this method to be generic as possible to support a wide range of Online Routing Engines.

Reply:
sure:
even though it's allowed to add a body to GET requests, it's commonly not done and [might cause issues with some existing implementations](https://www.rfc-editor.org/rfc/rfc7231#section-4.3.1).
I've just read [this](https://stackoverflow.com/questions/978061/http-get-with-request-body) to that topic.

Apart from POST and GET i don't think there will be other request methods to use for online routing engines.

Reply:
To avoid causing issues with existing implementations the `engine.getRequestBody` returns `null` by default. This should mitigate any risk of breaking those services.

The point of being generic is that we support things we cannot think of yet. Someone else can come in and add another online routing engine that behaves in a way we did not expect.

I am assuming you see the benefit of being generic. Do you see any risk with this implementation that might be mitigated by sending bodies only with POST requests?

Reply:
> Do you see any risk with this implementation that might be mitigated by sending bodies only with POST requests?

Not right now, as currently only one http method is allowed. In case an online routing engine (ORE) can support multiple methods in the future, this will be reworked anyway. At that point if the getRequestBody is defined in the ORE, but the GET request is executed, there might happen unexpected things.

I guess we can leave like that for now.

Reply:
Thanks for your thoughts :)